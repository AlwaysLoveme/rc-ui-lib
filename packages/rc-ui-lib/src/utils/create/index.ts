/* eslint-disable no-param-reassign */
import { BEM, createBEM } from './bem';

type CreateNamespaceReturn = [BEM, string];

export function createNamespace(name: string): CreateNamespaceReturn {
  name = `rc-${name}`;
  return [createBEM(name), name];
}
