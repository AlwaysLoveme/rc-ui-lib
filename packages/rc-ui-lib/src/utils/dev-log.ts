export function devWarning(component: string, message: string): void {
  if (process.env.NODE_ENV !== 'production') {
    console.warn(`[antd-mobile: ${component}] ${message}`);
  }
}

export function devError(component: string, message: string) {
  if (process.env.NODE_ENV !== 'production') {
    console.error(`[antd-mobile: ${component}] ${message}`);
  }
}
