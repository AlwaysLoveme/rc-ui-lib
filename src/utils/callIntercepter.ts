import { isPromise, noop } from './index';

export type Interceptor = (...args: any[]) => Promise<boolean> | boolean | void;

export const callIntercepter = (options: {
  interceptor?: Interceptor;
  args?: unknown[];
  done: () => void;
  canceled?: () => void;
}): any => {
  const { interceptor, args, done, canceled } = options;

  if (interceptor) {
    // eslint-disable-next-line prefer-spread
    const returnVal = interceptor.apply(null, args || []);

    if (isPromise(returnVal)) {
      returnVal
        .then((value) => {
          if (value) {
            done();
          } else if (canceled) {
            canceled();
          }
        })
        .catch(noop);
    } else if (returnVal) {
      done();
    } else if (canceled) {
      canceled();
    }
  } else {
    done();
  }
};
