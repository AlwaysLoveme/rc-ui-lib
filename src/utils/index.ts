import { FieldRule } from '../components/field/types';

const overflowScrollReg = /scroll|auto/i;
type ScrollElement = HTMLElement | Window;
export type ContanierType = HTMLElement | (() => HTMLElement) | Window;

export const noop = () => ({});

// 判断其参数是否是数字
export const isNumeric = (val: string): boolean => {
  return /^\d+(\.\d+)?$/.test(val);
};

// 判断其参数是否是非数字值
export const isNaN = (val: number): val is typeof NaN => {
  if (Number.isNaN) {
    return Number.isNaN(val);
  }
  return val !== val;
};

// 判断参数是否存在
export const isDef = <T>(val: T): val is NonNullable<T> => {
  return val !== undefined && val !== null;
};

// 添加单位
export const addUnit = (value?: string | number): string | undefined => {
  if (!isDef(value)) {
    return undefined;
  }
  value = String(value);
  return isNumeric(value) ? `${value}px` : value;
};
// 获取宽高
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const getSizeStyle = (originSize?: string | number) => {
  if (isDef(originSize)) {
    const size = addUnit(originSize);
    return {
      width: size,
      height: size,
    };
  }
  return {};
};
// 阻止默认事件
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const preventDefault = (event: MouseEvent, isStopPropagation?: boolean) => {
  if (typeof event.cancelable !== 'boolean' || event.cancelable) {
    event.preventDefault();
  }
  if (isStopPropagation) {
    event.stopPropagation();
  }
};
// 判断节点是否是元素
export const isElement = (node: Element): boolean => {
  const ELEMENT_NODE_TYPE = 1; // // 1：代表元素，详情可以搜索查看elementNode.nodeType值对应的类型。
  return node.tagName !== 'HTML' && node.tagName !== 'BODY' && node.nodeType === ELEMENT_NODE_TYPE;
};

export const getScrollParent = (el: Element, root: ScrollElement = window): Element | Window => {
  let node = el;
  while (node && node !== root && isElement(node)) {
    const { overflowY } = window.getComputedStyle(node);
    if (overflowScrollReg.test(overflowY)) {
      return node;
    }
    node = node.parentNode as Element;
  }
  return root;
};
// 判断是否是浏览器环境
export const inBrowser = (): boolean => {
  return typeof window !== 'undefined';
};

// 获取挂载的容器
export const getMountContanier = (contanier?: ContanierType): HTMLElement => {
  if (contanier) {
    if (typeof contanier === 'function') {
      return contanier();
    }
    if (typeof contanier === 'object' && contanier instanceof HTMLElement) {
      return contanier;
    }
  }
  return document.body;
};

// 将组件中传入的配置项(组件属性)对应的key/value赋给新对象。
export const pick = (prop: Record<string, any>, keys: string[]): Record<string, any> => {
  return keys.reduce((target, key) => {
    target[key] = prop[key];
    return target;
  }, {} as any);
};

export const isObject = (val: Record<string, any>): boolean => {
  return val !== null && typeof val === 'object';
};

// eslint-disable-next-line @typescript-eslint/ban-types
export const isFunction = (val: any): val is Function => {
  return val !== null && typeof val === 'function';
};

// 判断是否是Promise函数
export const isPromise = <T = any>(val: any): val is Promise<T> => {
  return isObject(val) && isFunction(val.then) && isFunction(val.catch);
};

export const isEmptyValue = (value: unknown): boolean => {
  if (Array.isArray(value)) {
    return !value.length;
  }
  if (value === 0) {
    return false;
  }
  return !value;
};

export function range(num: number, min: number, max: number): number {
  return Math.min(Math.max(num, min), max);
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const trimExtraChar = (value: string, char: string, regExp: RegExp) => {
  const index = value.indexOf(char);

  if (index === -1) {
    return value;
  }

  if (char === '-' && index !== 0) {
    return value.slice(0, index);
  }

  return value.slice(0, index + 1) + value.slice(index).replace(regExp, '');
};

export const formatNumber = (value: string, allowDot = true, allowMinus = true): string => {
  if (allowDot) {
    value = trimExtraChar(value, '.', /\./g);
  } else {
    value = value.split('.')[0];
  }
  if (allowMinus) {
    value = trimExtraChar(value, '-', /-/g);
  } else {
    value = value.replace(/-/, '');
  }
  const regExp = allowDot ? /[^-0-9.]/g : /[^-0-9]/g;
  return value.replace(regExp, '');
};

export const trigger = (target: Element, type: string): void => {
  const inputEvent = document.createEvent('HTMLEvents');
  inputEvent.initEvent(type, true, true);
  target.dispatchEvent(inputEvent);
};

export const runSyncRule = (value: unknown, rule: FieldRule): boolean => {
  if (rule.required && isEmptyValue(value)) {
    return false;
  }
  if (rule.pattern && !rule.pattern.test(String(value))) {
    return false;
  }
  return true;
};
