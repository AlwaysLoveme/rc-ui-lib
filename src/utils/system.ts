import { inBrowser } from './index';

export const isAndroid = (): boolean => {
  return inBrowser() ? /android/.test(navigator.userAgent.toLowerCase()) : false;
};

export const isIOS = (): boolean => {
  return inBrowser() ? /ios|iphone|ipad|ipod/.test(navigator.userAgent.toLowerCase()) : false;
};
