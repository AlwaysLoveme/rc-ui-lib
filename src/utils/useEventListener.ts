import { useEffect } from 'react';
import { noop } from './index';

export let supportsPassive = false;

if (typeof window !== 'undefined') {
    try {
        const opts = {};
        Object.defineProperty(opts, 'passive', {
            get() {
                supportsPassive = true;
                return true;
            }
        });
        window.addEventListener('test-passive', noop, opts);
    } catch (error) {}
}

export type useEventListenerOptions = {
    target?: EventTarget;
    capture?: boolean;
    passive?: boolean;
};

export const useEventListener = (
    type: string,
    listener: EventListener,
    options: useEventListenerOptions = {}
): any => {
    if (typeof window === 'undefined') {
        //  非浏览器环境
        return;
    }
    const { target = window, capture = false, passive = false } = options;

    let attached: boolean;

    const add = () => {
        if (target && !attached) {
            target.addEventListener(
                type,
                listener,
                supportsPassive ? { capture, passive } : capture
            );
            attached = true;
        }
    };

    const remove = () => {
        if (target && attached) {
            target.removeEventListener(type, listener, capture);
            attached = false;
        }
    };

    useEffect(() => {
        add();
        return () => {
            remove();
        };
    }, []);
};
