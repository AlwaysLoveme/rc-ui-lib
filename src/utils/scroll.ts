import { isIOS } from './system';

type ScrollElement = Element | Window;

const isWindow = (val: unknown): val is Window => {
  return val === window;
};

export const getScrollTop = (el: ScrollElement): number => {
  const top = 'scrollTop' in el ? el.scrollTop : el.pageYOffset;

  // iOS scroll bounce cause minus scrollTop
  return Math.max(top, 0);
};

export const setScrollTop = (el: ScrollElement, value: number): void => {
  if ('scrollTop' in el) {
    // eslint-disable-next-line no-param-reassign
    el.scrollTop = value;
  } else {
    el.scrollTo(el.scrollX, value);
  }
};

export const getRootScrollTop = (): number => {
  return window.pageYOffset || document.documentElement.scrollTop || document.body.scrollTop || 0;
};

export const setRootScrollTop = (value: number): void => {
  setScrollTop(window, value);
  setScrollTop(document.body, value);
};

// get distance from element top to page top or scroller top
export const getElementTop = (el: ScrollElement, scroller?: HTMLElement): number => {
  if (isWindow(el)) {
    return 0;
  }

  const scrollTop = scroller ? getScrollTop(scroller) : getRootScrollTop();
  return el.getBoundingClientRect().top + scrollTop;
};

export const getVisibleHeight = (el: ScrollElement): number => {
  if (isWindow(el)) {
    return el.innerHeight;
  }
  return el.getBoundingClientRect().height;
};

export const getVisibleTop = (el: ScrollElement): number => {
  if (isWindow(el)) {
    return 0;
  }
  return el.getBoundingClientRect().top;
};

// hack for iOS12 page scroll
// see: https://developers.weixin.qq.com/community/develop/doc/00044ae90742f8c82fb78fcae56800
export const resetScroll = (): void => {
  if (isIOS()) {
    setRootScrollTop(getRootScrollTop());
  }
};
