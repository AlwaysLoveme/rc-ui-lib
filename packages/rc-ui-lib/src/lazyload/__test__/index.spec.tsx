import React from 'react';
import { fireEvent, render, cleanup, waitFor } from '@testing-library/react';
import { Lazyload } from '..';
import { sleep } from '../../../tests/utils';

type ImageConstructor = new (
  width?: number | undefined,
  height?: number | undefined,
) => HTMLImageElement;

describe('test Lazyload by listen event', () => {
  global.Image = class {
    onload: () => void;

    constructor() {
      this.onload = jest.fn();
      setTimeout(() => {
        this.onload();
      }, 50);
    }
  } as unknown as ImageConstructor;
  const mockRect = (el: Element, rect: Partial<DOMRect>) => {
    Object.defineProperty(el, 'getBoundingClientRect', {
      configurable: true,
      get: () => {
        return () => ({
          top: 0,
          left: 0,
          right: 0,
          bottom: 0,
          width: 0,
          height: 0,
          ...rect,
        });
      },
    });
  };
  it('lazyload should render correctly when use listen event', async () => {
    const { container } = render(
      <Lazyload>
        <img alt="" src="https://img.yzcdn.cn/vant/apple-5.jpg" width="100%" height="300" />
      </Lazyload>,
    );
    expect(container).toMatchSnapshot();
  });
  it('lazyload should render correctly when load image failed', async () => {
    const { container } = render(
      <Lazyload>
        <img alt="" src="https://img.yzcdn.cn/vant/apple-5.jpg" width="100%" height="300" />
      </Lazyload>,
    );
    expect(container).toMatchSnapshot();
  });
  it('LazyloadImage should render correctly when use listen event', async () => {
    const { container } = render(<Lazyload.Image image="https://img.yzcdn.cn/vant/apple-5.jpg" />);
    expect(container).toMatchSnapshot();
  });

  it('LazyloadImage should emit onLoaded event when use listen event', async () => {
    const onLoaded = jest.fn();

    const { container, debug } = render(
      <Lazyload.Image onLoaded={onLoaded} image="https://img.yzcdn.cn/vant/apple-5.jpg" />,
    );
    const wrapper = container.querySelector('.rc-lazyload-image');

    mockRect(wrapper, {
      top: 10,
      left: 36,
      width: 300,
      height: 250,
    });
    mockRect(document.body, {
      width: 375,
      height: 800,
    });

    await waitFor(() => {
      fireEvent.scroll(window, { target: { pageYOffset: 30 } });
    });

    await sleep(1000);

    expect(onLoaded).toHaveBeenCalled();
  });
});
describe('test Lazyload by observer', () => {
  beforeEach(() => {
    // IntersectionObserver isn't available in test environment
    const mockIntersectionObserver = jest.fn();
    const observe = jest.fn();
    const unobserve = jest.fn();
    const disconnect = jest.fn();
    mockIntersectionObserver.mockReturnValue({
      observe,
      unobserve,
      disconnect,
    });
    window.IntersectionObserver = mockIntersectionObserver;
  });

  afterEach(() => {
    cleanup();
    jest.restoreAllMocks();
  });

  it('should render correctly when IntersectionObserver is available', async () => {
    const { container } = render(
      <Lazyload.Image image="https://img.yzcdn.cn/vant/apple-5.jpg" height="300px" />,
    );
    expect(container).toMatchSnapshot();
  });
  it('lazyload should render correctly when IntersectionObserver is available', async () => {
    const { container } = render(
      <Lazyload>
        <img alt="" src="https://img.yzcdn.cn/vant/apple-5.jpg" width="100%" height="300" />
      </Lazyload>,
    );
    expect(container).toMatchSnapshot();
  });
});
