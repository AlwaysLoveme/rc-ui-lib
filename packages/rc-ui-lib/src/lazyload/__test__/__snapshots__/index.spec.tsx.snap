// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Lazyload by listen event LazyloadImage should render correctly when use listen event 1`] = `
<div>
  <img
    class="rc-lazyload-image"
    data-src="https://img.yzcdn.cn/vant/apple-5.jpg"
    src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
  />
</div>
`;

exports[`test Lazyload by listen event lazyload should render correctly when use listen event 1`] = `
<div>
  <div
    class="rc-lazyload"
  >
    <div
      class="rc-lazyload__placeholder"
      style="height: 0px;"
    />
  </div>
</div>
`;

exports[`test Lazyload by observer lazyload should render correctly when IntersectionObserver is available 1`] = `
<div>
  <div
    class="rc-lazyload"
  >
    <div
      class="rc-lazyload__placeholder"
      style="height: 0px;"
    />
  </div>
</div>
`;

exports[`test Lazyload by observer should render correctly when IntersectionObserver is available 1`] = `
<div>
  <img
    class="rc-lazyload-image"
    data-src="https://img.yzcdn.cn/vant/apple-5.jpg"
    src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
    style="height: 300px;"
  />
</div>
`;
